src/main/java/com/ocient/jdbc/XGConnection.java

```
@@ -3653,61 +3658,79 @@ public void reconnect(final boolean shouldRequestVersion) throws IOException, SQ
+									connected = false;
 
-			throw new IOException("Failed to reconnect.");
-		}
+									return SQLResult.fail(successOrEx.getException());
+								});
+						}
+
+						return SQLResult.voidFuture();
+				});
 	}
 
-	private void clientHandshakeFinalize() throws SQLException, Exception {
+	private CompletionStage<SQLResult<Void>> clientHandshakeFinalize() {
 		// Set the default schema on the connection if one has not already been set
-		resetSchema().ifPresent(schema -> {
-			try
-			{
-				setSchema(schema);
-			}
-			catch (final Exception e)
-			{
-			}
-		});
 
-		if (setPso.isPresent()) {
-			if (setPso.getAsLong() == -1)
-			{
-				// We have to turn it off
-				setPSO(false);
-			}
-			else if (setPso.getAsLong() > 0)
-			{
-				// Set non-default threshold
-				setPSO(setPso.getAsLong());
-			}
-		}
+		return resetSchema()
+		    .map(schema -> setSchemaAsync(schema))
+		    .orElseGet(SQLResult::voidFuture)
+		    .thenCompose(schemaUpdated -> {
+		        if (setPso.isPresent()) {
+ 			        if (setPso.getAsLong() == -1)
+ 			        {
+ 				        // We have to turn it off
+ 				        try {
+							return setPSO(false);
+						} catch (Exception e) {
+							return CompletableFuture.failedFuture(e);
+						}
+ 			        }
+ 			        else if (setPso.getAsLong() > 0)
+ 			        {
+ 				        // Set non-default threshold
+ 				        try {
+							return setPSO(setPso.getAsLong());
+						} catch (Exception e) {
+							return CompletableFuture.failedFuture(e);
+						}
+ 			        }
+ 		        }
 
-		resendParameters();
+ 		        // nothing more to do
+ 		        return SQLResult.voidFuture();
+		    })
+			.thenCompose(successOrEx -> { 
+
+				if(successOrEx.isException()) {
+					SQLException e = successOrEx.getException();
+					LOGGER.log(Level.WARNING, String.format("Exception %s occurred in clientHandshakeFinalize() with message %s", e.toString(), e.getMessage()));
+				}
```

wondering if we actually want to swallow this exception, I'll have to look at the legacy impl. Sorry we don't have existing unit tests for stuff like this 🤦🏾